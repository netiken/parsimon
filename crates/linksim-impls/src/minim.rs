//! An interface to the Minim link-level simulator.

use std::{fs, io::Write, path::PathBuf, time::Instant};

use parsimon_core::{
    linksim::{LinkSim, LinkSimError, LinkSimResult},
    network::{Channel, EdgeIndex, FctRecord, FlowId, SimNetwork},
    units::{BitsPerSec, Bytes, Kilobytes, Nanosecs},
};
use rustc_hash::{FxHashMap, FxHashSet};

use crate::utils;

// XXX: These are set to match the ns3 implementation's default behavior.
const SZ_PKTMAX: Bytes = Bytes::new(1_000);
const SZ_PKTHDR: Bytes = Bytes::new(48);

/// An Minim link simulation.
#[derive(Debug, typed_builder::TypedBuilder)]
pub struct MinimLink {
    #[builder(setter(into))]
    window: Bytes,
    dctcp_gain: f64,
    #[builder(setter(into))]
    dctcp_ai: BitsPerSec,

    // Stats
    #[builder(default, setter(strip_option))]
    elapsed_record: Option<PathBuf>,
}

impl LinkSim for MinimLink {
    fn simulate(&self, network: &SimNetwork, edge: EdgeIndex) -> LinkSimResult {
        let chan = network
            .edge(edge)
            .ok_or_else(|| LinkSimError::UnknownEdge(edge))?;
        let flows = network.flows_on(edge).unwrap(); // we already know the channel exists

        let src_map = flows.iter().map(|f| f.src).collect::<FxHashSet<_>>();
        let (bsrc, bdst) = (chan.src(), chan.dst());

        let srcs = src_map
            .iter()
            .map(|&src| {
                let (delay2btl, link_rate) = if src == bsrc {
                    let path = network.path(src, bdst, |c| c.first());
                    let &(eidx, chan) = path.iter().next().unwrap();
                    let link_rate = chan.bandwidth() - utils::ack_rate(network, eidx);
                    (Nanosecs::ZERO, link_rate)
                } else {
                    let path = network.path(src, bsrc, |c| c.first());
                    let &(eidx, chan) = path.iter().next().unwrap();
                    let link_rate = chan.bandwidth() - utils::ack_rate(network, eidx);
                    (path.delay(), link_rate)
                };
                minim::SourceDesc::builder()
                    .id(minim::SourceId::new(src.inner()))
                    .delay2btl(minim::units::Nanosecs::new(delay2btl.into_u64()))
                    .link_rate(minim::units::BitsPerSec::new(link_rate.into_u64()))
                    .build()
            })
            .collect::<Vec<_>>();

        let mut src2dst2delay = FxHashMap::default();
        let flows = flows
            .into_iter()
            .map(|f| {
                let delay2dst = *src2dst2delay
                    .entry(f.src)
                    .or_insert_with(FxHashMap::default)
                    .entry(f.dst)
                    .or_insert_with(|| {
                        let path = network.path(f.src, f.dst, |c| c.first());
                        path.delay()
                    });
                minim::FlowDesc {
                    id: minim::FlowId::new(f.id.inner()),
                    source: minim::SourceId::new(f.src.inner()),
                    size: minim::units::Bytes::new(f.size.into_u64()),
                    start: minim::units::Nanosecs::new(f.start.into_u64()),
                    delay2dst: minim::units::Nanosecs::new(delay2dst.into_u64()),
                }
            })
            .collect::<Vec<_>>();

        let marking_threshold = Kilobytes::new(
            chan.bandwidth()
                .scale_by(1e9_f64.recip())
                .scale_by(3_f64)
                .into_u64(),
        );
        let bandwidth = if src_map.contains(&chan.src()) {
            chan.bandwidth().scale_by(100_f64)
        } else {
            chan.bandwidth() - utils::ack_rate(network, edge)
        };
        let cfg = minim::Config::builder()
            .bandwidth(minim::units::BitsPerSec::new(bandwidth.into_u64()))
            .queue(minim::queue::FifoQ::new())
            .sources(srcs)
            .flows(flows)
            .window(minim::units::Bytes::new(self.window.into_u64()))
            .dctcp_marking_threshold(minim::units::Kilobytes::new(marking_threshold.into_u64()))
            .dctcp_gain(self.dctcp_gain)
            .dctcp_ai(minim::units::BitsPerSec::new(self.dctcp_ai.into_u64()))
            .sz_pktmax(minim::units::Bytes::new(SZ_PKTMAX.into_u64()))
            .sz_pkthdr(minim::units::Bytes::new(SZ_PKTHDR.into_u64()))
            .build();

        let start = Instant::now();
        let records = minim::run(cfg);
        let elapsed_millis = start.elapsed().as_millis();
        if let Some(path) = &self.elapsed_record {
            let mut file = fs::OpenOptions::new()
                .append(true)
                .create(true)
                .open(path)?;
            file.write_all(format!("{}\n", elapsed_millis).as_bytes())?;
        }

        let records = records
            .into_iter()
            .map(|r| FctRecord {
                id: FlowId::new(r.id.into_usize()),
                size: Bytes::new(r.size.into_u64()),
                start: Nanosecs::new(r.start.into_u64()),
                fct: Nanosecs::new(r.fct.into_u64()),
                ideal: Nanosecs::new(r.ideal.into_u64()),
            })
            .collect();

        Ok(records)
    }
}
