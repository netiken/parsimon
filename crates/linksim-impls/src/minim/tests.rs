use std::collections::BTreeMap;

use minim::{
    queue::FifoQ,
    units::{BitsPerSec, Bytes, Nanosecs},
    Config, FlowDesc, SourceDesc,
};
use parsimon_core::{
    network::{Channel, Flow, FlowId, Network, NodeId},
    testing,
};
use rand::prelude::*;
use rand_distr::Exp;

use super::MinimLink;

#[derive(Debug, serde::Serialize)]
struct MinimCheck {
    bandwidth: BitsPerSec,
    sources: Vec<SourceDesc>,
    flows: Vec<FlowDesc>,
    window: Bytes,
    dctcp_marking_threshold: Bytes,
    dctcp_gain: f64,
    dctcp_ai: BitsPerSec,
    sz_pktmax: Bytes,
    sz_pkthdr: Bytes,
    timeout: Option<Nanosecs>,
}

impl MinimCheck {
    fn from_config(cfg: &Config<FifoQ>) -> Self {
        Self {
            bandwidth: cfg.bandwidth,
            sources: cfg.sources.clone(),
            flows: cfg.flows.clone(),
            window: cfg.window,
            dctcp_marking_threshold: cfg.dctcp_marking_threshold,
            dctcp_gain: cfg.dctcp_gain,
            dctcp_ai: cfg.dctcp_ai,
            sz_pktmax: cfg.sz_pktmax,
            sz_pkthdr: cfg.sz_pkthdr,
            timeout: cfg.timeout,
        }
    }
}

type Snapshot = BTreeMap<(NodeId, NodeId), MinimCheck>;

fn eight_node_config_snapshots(flows: Vec<Flow>) -> anyhow::Result<Snapshot> {
    // Build a `Network`.
    let (nodes, links) = testing::eight_node_config();
    let network = Network::new(&nodes, &links)?;

    // Convert the `Network` into a `SimNetwork` by adding flows.
    let network = network.into_simulations(flows);

    // Build a `MinimLink` instance and use it to generate `MinimCheck`s.
    let linksim = MinimLink::builder()
        .window(parsimon_core::units::Bytes::new(18_000))
        .dctcp_gain(0.0625)
        .dctcp_ai(parsimon_core::units::Mbps::new(615))
        .build();
    let snapshot = network
        .edge_indices()
        .map(|eidx| {
            let cfg = linksim.build_config(&network, eidx)?;
            let chan = network
                .edge(eidx)
                .ok_or_else(|| anyhow::anyhow!("unknown edge"))?;
            let check = MinimCheck::from_config(&cfg);
            Ok(((chan.src(), chan.dst()), check))
        })
        .collect::<anyhow::Result<BTreeMap<(NodeId, NodeId), MinimCheck>>>()?;
    Ok(snapshot)
}

fn gen_flows(
    mean_flow_size: parsimon_core::units::Bytes,
    mean_load: f64,
    nr_flows: usize,
    mut rng: impl Rng,
) -> anyhow::Result<Vec<Flow>> {
    // Calculate mean interarrival time T (ns) for one server
    // Bandwidth (bps) * Load (bps/bps) = desired rate (bps)
    // flow size (bytes) to flow size (bits) / desired rate (bps) --> seconds --> ns
    let bandwidth_bps = parsimon_core::units::BitsPerSec::from(parsimon_core::units::Gbps::new(10));
    let desired_rate = bandwidth_bps.into_f64() * mean_load;
    let mean_interarrival_time = parsimon_core::units::Nanosecs::new(
        ((mean_flow_size.into_f64() * 8.0 * 1e9) / (desired_rate * 4.0)) as u64,
    );

    // Make exponential distributions
    // Flow size distribution
    let flow_exp = Exp::new(mean_flow_size.into_f64().recip())?;
    let start_exp = Exp::new(mean_interarrival_time.into_f64().recip())?;
    let mut node_nums: Vec<usize> = (0..4).collect();

    // Draw flows from distribution
    let mut flows: Vec<Flow> = Vec::new();
    let mut prev_start: u64 = 0;
    for i in 0..nr_flows {
        node_nums.shuffle(&mut rng);
        let new_start: u64 = start_exp.sample(&mut rng).round() as u64 + prev_start;
        flows.push(Flow {
            id: FlowId::new(i),
            src: NodeId::new(node_nums[0]),
            dst: NodeId::new(node_nums[1]),
            size: parsimon_core::units::Bytes::new(flow_exp.sample(&mut rng).round() as u64),
            start: parsimon_core::units::Nanosecs::new(new_start),
        });
        prev_start = new_start;
    }
    Ok(flows)
}

#[test]
fn config_correct() -> anyhow::Result<()> {
    let snapshot = eight_node_config_snapshots(vec![
        Flow {
            id: FlowId::ZERO,
            src: NodeId::new(0),
            dst: NodeId::new(2),
            size: parsimon_core::units::Bytes::new(1000),
            start: parsimon_core::units::Nanosecs::ZERO,
        },
        Flow {
            id: FlowId::ONE,
            src: NodeId::new(0),
            dst: NodeId::new(2),
            size: parsimon_core::units::Bytes::new(1000),
            start: parsimon_core::units::Nanosecs::new(960),
        },
    ])?;
    insta::assert_yaml_snapshot!(snapshot);
    Ok(())
}

#[test]
fn config_correct_loaded() -> anyhow::Result<()> {
    let rng = StdRng::seed_from_u64(0);
    let flows = gen_flows(parsimon_core::units::Bytes::new(10_000), 0.5, 10000, rng)?;
    let snapshot = eight_node_config_snapshots(flows)?;
    insta::assert_yaml_snapshot!(snapshot);
    Ok(())
}
