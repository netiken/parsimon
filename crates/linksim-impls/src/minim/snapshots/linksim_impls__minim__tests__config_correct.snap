---
source: crates/linksim-impls/src/minim/tests.rs
expression: snapshot
---
? - 0
  - 4
: bandwidth: 1000000000000
  sources:
    - id: 0
      delay2btl: 0
      link_rate: 10000000000
  flows:
    - id: 0
      source: 0
      qindex: 0
      size: 1000
      start: 0
      delay2dst: 4000
    - id: 1
      source: 0
      qindex: 0
      size: 1000
      start: 960
      delay2dst: 4000
  window: 18000
  dctcp_marking_threshold: 30000
  dctcp_gain: 0.0625
  dctcp_ai: 615000000
  sz_pktmax: 1000
  sz_pkthdr: 48
  timeout: ~
? - 4
  - 6
: bandwidth: 10000000000
  sources:
    - id: 0
      delay2btl: 1000
      link_rate: 10000000000
  flows:
    - id: 0
      source: 0
      qindex: 0
      size: 1000
      start: 0
      delay2dst: 4000
    - id: 1
      source: 0
      qindex: 0
      size: 1000
      start: 960
      delay2dst: 4000
  window: 18000
  dctcp_marking_threshold: 30000
  dctcp_gain: 0.0625
  dctcp_ai: 615000000
  sz_pktmax: 1000
  sz_pkthdr: 48
  timeout: ~
? - 5
  - 2
: bandwidth: 10000000000
  sources:
    - id: 0
      delay2btl: 3000
      link_rate: 10000000000
  flows:
    - id: 0
      source: 0
      qindex: 0
      size: 1000
      start: 0
      delay2dst: 4000
    - id: 1
      source: 0
      qindex: 0
      size: 1000
      start: 960
      delay2dst: 4000
  window: 18000
  dctcp_marking_threshold: 30000
  dctcp_gain: 0.0625
  dctcp_ai: 615000000
  sz_pktmax: 1000
  sz_pkthdr: 48
  timeout: ~
? - 6
  - 5
: bandwidth: 10000000000
  sources:
    - id: 0
      delay2btl: 2000
      link_rate: 10000000000
  flows:
    - id: 0
      source: 0
      qindex: 0
      size: 1000
      start: 0
      delay2dst: 4000
    - id: 1
      source: 0
      qindex: 0
      size: 1000
      start: 960
      delay2dst: 4000
  window: 18000
  dctcp_marking_threshold: 30000
  dctcp_gain: 0.0625
  dctcp_ai: 615000000
  sz_pktmax: 1000
  sz_pkthdr: 48
  timeout: ~
